src/dbnode/persist/fs/write.go

```diff
@@ -327,13 +329,38 @@ func (w *writer) Close() error {
 	}
 	// NB(xichen): only write out the checkpoint file if there are no errors
 	// encountered between calling writer.Open() and writer.Close().
-	if err := w.writeCheckpointFile(); err != nil {
+	if err := w.writeCheckpointFile(
+		w.checkpointFilePath,
+		w.digestFdWithDigestContents.Digest().Sum32(),
+		w.digestBuf,
+	); err != nil {
 		w.err = err
 		return err
 	}
 	return nil
 }
 
+func (w *writer) DeferClose() (persist.DataCloser, error) {
+	err := w.close()
+	if w.err != nil {
+		return nil, w.err
+	}
+	if err != nil {
+		w.err = err
+		return nil, err
+	}
+	checkpointFilePath := w.checkpointFilePath
+	digestBuf := w.digestBuf
+	digestChecksum := w.digestFdWithDigestContents.Digest().Sum32()
+	return func() error {
+		return w.writeCheckpointFile(
```

Comment:
Are there any synchronization concerns here? May be worth adding a comment to the interface about usage?

Reply:
This whole structure is meant to be called in a single threaded way.

I do think though it is reused, so perhaps we shouldn't be using the the `w.digestBuf` (which seems to be reused?)

Might be best to even just to do entirely inline without shared resources and create a new stateless "func writeCheckpointFile(path string, checksum uint32) error" like so:

```go
 func writeCheckpointFile(filePath string, checksum uint32, newFileMode os.FileMode) error {
  fd, err := OpenWritable(filePath, newFileMode)
  // handle err ...
  err := digest.NewBuffer().WriteDigestToFile(fd, checksum)
  // handle err ...
  return fd.Close()
}
```